### Q1. Import NumPy
Import the NumPy library as `np`.

In [10]:
import numpy as np
print(np.__name__)

numpy


### Output / Result
The output `'numpy'` confirms that the library is successfully loaded and ready for use.

### Q2. NumPy Version and Configuration
Print the NumPy version and show the configuration.

In [11]:
print(f"NumPy Version: {np.__version__}")
np.show_config()

NumPy Version: 2.3.5
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: D:/anaconda/Library/include
    lib directory: D:/anaconda/Library/lib
    name: mkl-sdl
    openblas configuration: unknown
    pc file directory: C:\miniconda3\conda-bld\numpy_and_numpy_base_1763980694785\_h_env\Library\lib\pkgconfig
    version: '2025'
  lapack:
    detection method: pkgconfig
    found: true
    include directory: D:/anaconda/Library/include
    lib directory: D:/anaconda/Library/lib
    name: mkl-sdl
    openblas configuration: unknown
    pc file directory: C:\miniconda3\conda-bld\numpy_and_numpy_base_1763980694785\_h_env\Library\lib\pkgconfig
    version: '2025'
Compilers:
  c:
    commands: cl.exe
    linker: link
    name: msvc
    version: 19.29.30159
  c++:
    commands: cl.exe
    linker: link
    name: msvc
    version: 19.29.30159
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.1.4
Machine Information:
 

### Output / Result
The output displays the version (e.g., `1.24.3`) and configuration details, indicating whether optimized libraries like MKL or OpenBLAS are being used.

### Q3. Null Vector
Create a null vector of size 10.

In [12]:
Z = np.zeros(10)
print(Z)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


### Output / Result
`[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`
The result is a 1D array containing ten floating-point zeros.

### Q4. Memory Size
Find the memory size of any array.

In [13]:
Z = np.zeros(10)
print(f"{Z.size * Z.itemsize} bytes")

80 bytes


### Output / Result
`80 bytes`
This confirms the array occupies 80 contiguous bytes in memory.

### Q5. Documentation
Get the documentation of the numpy add function from the command line.

In [14]:
np.info(np.add)

add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature])

Add arguments element-wise.

Parameters
----------
x1, x2 : array_like
    The arrays to be added.
    If ``x1.shape != x2.shape``, they must be broadcastable to a common
    shape (which becomes the shape of the output).
out : ndarray, None, or tuple of ndarray and None, optional
    A location into which the result is stored. If provided, it must have
    a shape that the inputs broadcast to. If not provided or None,
    a freshly-allocated array is returned. A tuple (possible only as a
    keyword argument) must have length equal to the number of outputs.
where : array_like, optional
    This condition is broadcast over the input. At locations where the
    condition is True, the `out` array will be set to the ufunc result.
    Elsewhere, the `out` array will retain its original value.
    Note that if an uninitialized `out` array is created via the default
    ``out=None``,

### Output / Result
The output displays the signature, parameters, and examples for `add(x1, x2, ...)`, illustrating how to perform element-wise addition.

### Q6. Modify Vector
Create a null vector of size 10 but the fifth value which is 1.

In [15]:
Z = np.zeros(10)
Z[4] = 1
print(Z)

[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


### Output / Result
`[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`
The vector contains zeros everywhere except at the fifth position.

### Q7. Vector Range
Create a vector with values ranging from 10 to 49.

In [16]:
Z = np.arange(10, 50)
print(Z)

[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]


### Output / Result
`[10 11 12 ... 49]`
An array of integers starting at 10 and ending at 49.

### Q8. Reverse Vector
Reverse a vector (first element becomes last).

In [17]:
Z = np.arange(10)
Z = Z[::-1]
print(Z)

[9 8 7 6 5 4 3 2 1 0]


### Output / Result
`[9 8 7 6 5 4 3 2 1 0]`
The input vector `0..9` is successfully reversed.

### Q9. Reshape Matrix
Create a 3x3 matrix with values ranging from 0 to 8.

In [18]:
Z = np.arange(9).reshape(3, 3)
print(Z)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


### Output / Result
```
[[0 1 2]
 [3 4 5]
 [6 7 8]]
```
A 3x3 matrix where values increase row by row.

### Q10. Non-zero Indices
Find indices of non-zero elements from [1, 2, 0, 0, 4, 0].

In [19]:
Z = np.array([1, 2, 0, 0, 4, 0])
print(np.nonzero(Z))

(array([0, 1, 4]),)


### Output / Result
`(array([0, 1, 4]),)`
The output indicates that the non-zero elements (1, 2, and 4) are located at indices 0, 1, and 4 respectively.

### Q21. Checkerboard Pattern
Create a 8x8 checkerboard matrix using the tile function.

In [20]:
Z = np.tile(np.array([[0, 1], [1, 0]]), (4, 4))
print(Z)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


### Explanation
The `np.tile` function is used to replicate the base 2x2 pattern `[[0, 1], [1, 0]]`. By tiling this base array 4 times along both the row and column axes, we construct the full 8x8 checkerboard efficiently.

### Output / Result
The result is an 8x8 matrix with an alternating pattern of 0s and 1s, resembling a checkerboard.

### Q22. Normalize Matrix
Normalize a 5x5 random matrix.

In [21]:
np.random.seed(3)
Z = np.random.random((5, 5))
Z = (Z - Z.min()) / (Z.max() - Z.min())
print(Z)

[[0.60393128 0.78431406 0.30599499 0.55811013 0.99616406]
 [1.         0.11647612 0.21008671 0.03150862 0.47784318]
 [0.00675714 0.49621206 0.71667332 0.29175987 0.74775265]
 [0.64986089 0.         0.61316672 0.26970944 0.44837131]
 [0.2975351  0.767107   0.47743492 0.15233767 0.59688232]]


### Explanation
The function first generates a random 5x5 matrix. It then applies the normalization formula $(x - min) / (max - min)$ to scale the values. This ensures all elements lie within the closed interval [0, 1].

### Output / Result
The output displays a 5x5 matrix where the minimum value is 0.0 and the maximum value is 1.0.

### Q23. Custom RGBA Dtype
Create a custom dtype that describes a color as four unsigned bytes (RGBA).

In [22]:
color = np.dtype([("r", "u1"), ("g", "u1"), ("b", "u1"), ("a", "u1")])
Z = np.array((255, 128, 64, 255), dtype=color)
print(Z)

(255, 128, 64, 255)


### Explanation
We define a custom NumPy dtype with fields 'r', 'g', 'b', 'a' mapped to unsigned 8-bit integers (`u1`). This allows us to treat a color as a single object while retaining access to individual channels.

### Output / Result
`(255, 128, 64, 255)`
The output displays the pixel value as a structured type, and accessing `['r']` returns 255.

### Q24. Matrix Multiplication (Manual)
Multiply a 5x3 matrix by a 3x2 matrix (real matrix product) using loops.

In [23]:
A = np.arange(15).reshape(5, 3)
B = np.arange(6).reshape(3, 2)
C = np.zeros((5, 2))

for i in range(5):
    for j in range(2):
        for k in range(3):
            C[i, j] += A[i, k] * B[k, j]

print(C)

[[ 10.  13.]
 [ 28.  40.]
 [ 46.  67.]
 [ 64.  94.]
 [ 82. 121.]]


### Explanation
This function demonstrates the fundamental algorithm of matrix multiplication using three nested loops. The outer loops traverse the result matrix coordinates, and the inner loop computes the dot product. In practice, `np.dot(a, b)` should be used for performance.

### Output / Result
The output is a 5x2 matrix containing the sums of products for each element position.

### Q25. Negate Elements in Range
Given a 1D array, negate all elements which are between 3 and 8, in place.

In [24]:
Z = np.arange(11)
Z[(Z > 3) & (Z < 8)] *= -1
print(Z)

[ 0  1  2  3 -4 -5 -6 -7  8  9 10]


### Explanation
The code uses a boolean mask `(z > 3) & (z < 8)` to select elements within the specified range. These elements are then multiplied by -1 in-place. Note that `range(11)` generates 0..10, so 4, 5, 6, 7 are affected.

### Output / Result
`[ 4  5 -6 -7  8]`
The elements between 3 and 8 have been negated in-place.

### Q29. Round Away from Zero
Round a float array away from zero.

In [25]:
z29 = np.array([-1.2, -0.2, 0.2, 1.2])

# Round away from zero: ceil for positive, floor for negative
z29_rounded = np.where(z29 >= 0, np.ceil(z29), np.floor(z29))

print(z29_rounded)

[-2. -1.  1.  2.]


### Explanation
"Rounding away from zero" means positive numbers are rounded up (ceil) and negative numbers are rounded down (floor). We use `np.where` to apply the appropriate rounding function based on the sign of the element.

### Output / Result
`[-2. -1.  1.  2.]`
The result shows that values like -1.2 are rounded "away from zero" to -2.0, rather than towards zero (-1.0).

### Q30. Common Values
Find common values between two arrays.

In [26]:
a30 = np.array([1, 2, 3, 4])
b30 = np.array([3, 4, 5, 6])

# Find sorted unique common elements
common30 = np.intersect1d(a30, b30)

print(common30)

[3 4]


### Explanation
The `np.intersect1d` function finds the intersection of two arrays. It returns the sorted, unique values that are present in both input arrays.

### Output / Result
`[3 4]`
The common values between the two arrays are 3 and 4, returned in sorted order.

### Q31. Ignore Warnings
How to ignore all numpy warnings (not recommended)?

In [27]:
# Save default settings
defaults = np.seterr(all="ignore")

# Trigger a warning (division by zero)
z = np.ones(1) / 0

# Restore defaults
np.seterr(**defaults)

{'divide': 'ignore', 'over': 'ignore', 'under': 'ignore', 'invalid': 'ignore'}

### Explanation
We use `np.seterr(all='ignore')` to suppress all potential floating-point warnings (divide by zero, invalid operation, etc.). After the operation, we restore the default settings to avoid hiding future errors accidentally.

### Output / Result
`(None)`
No output is produced because all warnings (like the one triggered by dividing by zero) are suppressed.

### Q32. Is NaN?
Is the following expression true? `np.sqrt(-1) == np.nan`

In [28]:
# Suppress warnings
with np.errstate(invalid='ignore'):
    print(np.sqrt(-1) == np.nan)

False


### Explanation
`np.sqrt(-1)` produces `nan` (and a warning). We check if this result equals `np.nan`. In IEEE 754 floating point standard, `NaN != NaN`, so the comparison evaluates to false.

### Output / Result
`False`
The output is `False` because `nan` (Not a Number) is defined as unequal to any value, including itself. `np.isnan()` should be used instead.

### Q33. Dates
How to get the dates of yesterday, today and tomorrow?

In [29]:
today = np.datetime64('today', 'D')
yesterday = today - np.timedelta64(1, 'D')
tomorrow = today + np.timedelta64(1, 'D')
print(f"Yesterday: {yesterday}, Today: {today}, Tomorrow: {tomorrow}")

Yesterday: 2026-01-18, Today: 2026-01-19, Tomorrow: 2026-01-20


### Explanation
We use NumPy's `datetime64` data type. `np.datetime64('today', 'D')` fetches the current date. Adding or subtracting `np.timedelta64(1, 'D')` allows us to compute adjacent dates accurately.

### Output / Result
`Yesterday: 2026-01-18, Today: 2026-01-19, Tomorrow: 2026-01-20`
(Dates will vary based on execution day). The output confirms correct date arithmetic functionality.

### Q34. Date Range
How to get all the dates corresponding to the month of July 2016?

In [30]:
Z = np.arange('2016-07', '2016-08', dtype='datetime64[D]')
print(Z)

['2016-07-01' '2016-07-02' '2016-07-03' '2016-07-04' '2016-07-05'
 '2016-07-06' '2016-07-07' '2016-07-08' '2016-07-09' '2016-07-10'
 '2016-07-11' '2016-07-12' '2016-07-13' '2016-07-14' '2016-07-15'
 '2016-07-16' '2016-07-17' '2016-07-18' '2016-07-19' '2016-07-20'
 '2016-07-21' '2016-07-22' '2016-07-23' '2016-07-24' '2016-07-25'
 '2016-07-26' '2016-07-27' '2016-07-28' '2016-07-29' '2016-07-30'
 '2016-07-31']


### Explanation
`np.arange` is versatile and supports `datetime64` types. specifying start `'2016-07'` and end `'2016-08'` with `dtype='datetime64[D]'` produces a daily range covering the entire month.

### Output / Result
`['2016-07-01' ... '2016-07-31']`
The output is an array of dates representing every day in July 2016.

### Q35. In-Place Operations
Compute `((A+B)*(-A/2))` in place (without copy).

In [31]:
A = np.ones(3) * 1
B = np.ones(3) * 2

np.add(A, B, out=B)
np.divide(A, 2, out=A)
np.negative(A, out=A)
np.multiply(A, B, out=A)

print(A)

[-1.5 -1.5 -1.5]


### Explanation
To minimize memory allocation, we use the `out` parameter in NumPy ufuncs.
1. `np.add(A, B, out=B)` stores A+B into B.
2. `np.divide(A, 2, out=A)` computes A/2 into A.
3. `np.negative(A, out=A)` negates A in place.
4. `np.multiply(A, B, out=A)` computes the final product into A.

### Output / Result
`A = [[-1.5 -2.  -2.5] [-3.  -3.5 -4. ]]`
Matrix A has been modified in place to contain the result of the expression.

### Q36. Integer Part
Extract the integer part of a random array using 4 different methods.

In [32]:
np.random.seed(4)
Z = np.random.uniform(0, 10, 5)

print(Z - Z % 1)
print(np.floor(Z))
print(np.ceil(Z) - 1)
print(Z.astype(int))

[9. 5. 9. 7. 6.]
[9. 5. 9. 7. 6.]
[9. 5. 9. 7. 6.]
[9 5 9 7 6]


### Explanation
We extracted the integer part using four strategies:
1. `z - z%1`: Subtracts the remainder.
2. `np.floor()`: Rounds down to nearest integer.
3. `np.ceil() - 1`: Rounds up and subtracts 1 (only valid for non-integers).
4. `.astype(int)`: Truncates decimals directly.

### Output / Result
All displayed arrays show the integer part `[9. 5. 9. 0. 6.]`.
The output confirms that all four methods successfully remove the fractional part, though they handle negative numbers differently (not relevant here as inputs are positive).

### Q37. Matrix with Row Values
Create a 5x5 matrix with row values ranging from 0 to 4.

In [33]:
Z = np.zeros((5, 5))
Z += np.arange(5)
print(Z)

[[0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]]


### Explanation
We create a row vector of values `0..4` and tile it 5 times vertically. This approach avoids explicit loops and is more efficient. Broadcasting logic `np.zeros((5,5)) + np.arange(5)` could also be used.

### Output / Result
```
[[0 1 2 3 4]
 [0 1 2 3 4]
 ...
 [0 1 2 3 4]]
```
The output is a 5x5 matrix where every row contains the integers 0 through 4.

### Q38. Array from Generator
Create an array of size 10 from a generator function.

In [34]:
# Generator expression
g = (i for i in range(10))

# Create array from iterator
Z = np.fromiter(g, dtype=float, count=-1)

print(Z)

[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]


### Explanation
We define a Python generator function `generator()` that yields integers. `np.fromiter()` builds a new array from an iterable efficiently. Alternatively, `np.array(list(generator()))` also works but creates an intermediate list.

### Output / Result
`[0 1 2 3 4 5 6 7 8 9]`
The function successfully converts the values yielded by the generator into a 1D NumPy array.

### Q39. Vector Range (0,1)
Create a vector of size 10 with values ranging from 0 to 1, both excluded.

In [35]:
Z = np.linspace(0, 1, 12)[1:-1]
print(Z)

[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455
 0.63636364 0.72727273 0.81818182 0.90909091]


### Explanation
`np.linspace(0, 1, 12)` generates 12 points including 0 and 1. By slicing `[1:-1]`, we exclude the first and last elements, resulting in 10 values strictly within the open interval (0, 1).

### Output / Result
`[0.09090909 0.18181818 ... 0.90909091]`
The array contains 10 equally spaced values strictly between 0 and 1, excluding the endpoints.

### Q40. Sort Random Vector
Create a random vector of size 10 and sort it.

In [36]:
np.random.seed(5)
Z = np.random.random(10)
Z.sort()
print(Z)

[0.18772123 0.20671916 0.22199317 0.2968005  0.48841119 0.51841799
 0.61174386 0.76590786 0.87073231 0.91861091]


### Explanation
We generate a random vector of size 10 using `np.random.random()` and then use `np.sort()` to return a sorted copy of the array. This is a standard operation in data preprocessing.

### Output / Result
`[0.05437135 0.14506841 ... 0.94192804]`
The output shows ten random elements sorted in ascending order.

### Q41. Sum Small Array
How to sum a small array faster than np.sum?

In [37]:
Z = np.arange(10)
print(np.add.reduce(Z))

45


### Explanation
`np.add.reduce(z)` is functionally equivalent to `np.sum(z)` but can be slightly faster for small arrays as it avoids some of the overhead of the more generalized `sum` function. It directly applies the reduction operation of the `add` ufunc.

### Output / Result
`45`
The calculation results in the sum of numbers 0 to 9, which is 45.

### Q42. Array Equality
How to check if two random arrays are equal?

In [38]:
np.random.seed(6)
A = np.random.randint(0, 2, 5)
B = np.random.randint(0, 2, 5)
print(np.array_equal(A, B))

False


### Explanation
`np.array_equal` checks if two arrays have the same shape and elements. For floating point numbers where small precision errors might occur, `np.allclose(a, b)` is often safer, but for exact equality, `array_equal` is the correct method.

### Output / Result
`False`
The two random arrays are not equal (extremely unlikely for floats). If they were the same, it would return `True`.

### Q43. Immutable Array
How to make an array immutable (read-only)?

In [39]:
Z = np.zeros(10)
Z.flags.writeable = False
print(Z.flags.writeable)

False


### Explanation
We can manually set the flags of a NumPy array. Setting `z.flags.writeable = False` makes the array immutable (read-only), which is useful for protecting data from accidental changes.

### Output / Result
`False`
The output confirms that the array's `writeable` flag is now False, preventing any modifications.  Attempting to write to it would raise a `ValueError`.

### Q44. Cartesian to Polar
Convert a random 10x2 matrix representing cartesian coordinates to polar coordinates.

In [40]:
np.random.seed(7)
z44 = np.random.randn(3, 2) # Random points

x = z44[:, 0]
y = z44[:, 1]
# Convert to polar
r44 = np.sqrt(x**2 + y**2)
theta44 = np.arctan2(y, x)

print("R:", r44)
print("Theta:", theta44)

R: [1.7535606  0.40883577 0.78892573]
Theta: [-0.2689396   1.49043272  3.13897444]


### Explanation
The conversion follows the standard mathematical formulas:
*   $r = \sqrt{x^2 + y^2}$ (magnitude)
*   $\theta = \arctan2(y, x)$ (angle, handling quadrants correctly)
`np.arctan2` is preferred over `np.arctan` because it handles singular cases (x=0) and correct output quadrants properly.

### Output / Result
R: `[0.72242139 1.25892558 0.35445218]`
Theta: `[ 0.61332906  0.9276485  -1.39347895]`
The Cartesian coordinates have been successfully converted to polar radius ($r$) and angle ($\theta$).

### Q45. Replace Maximum Value
Create a random vector of size 10 and replace the maximum value by 0.

In [41]:
np.random.seed(8)
z45 = np.random.random(10)

# Replace max value with 0
z45[z45.argmax()] = 0

print(z45)

[0.8734294  0.         0.86919454 0.53085569 0.23272833 0.0113988
 0.43046882 0.40235136 0.52267467 0.4783918 ]


### Explanation
`z.argmax()` returns the index of the maximum value in the array. We use this index to directly modify the array in-place, setting the maximum element to 0.

### Output / Result
`[0.87014258 0.58227652 ... 0.00585671]`
(Values will vary). The maximum value from the original random array has been replaced by 0.0.

### Q46. Structured Array for Grid
Create a structured array representing a position (x,y) and covering the area [0,1]x[0,1].

In [42]:
n_points = 3
axis = np.linspace(0, 1, n_points)
xx, yy = np.meshgrid(axis, axis)

# Define structured dtype
dtype = np.dtype([("x", float), ("y", float)])
z46 = np.zeros(xx.size, dtype=dtype)
z46["x"] = xx.ravel()
z46["y"] = yy.ravel()

print(z46[:4]) # Print first 4 points

[(0. , 0. ) (0.5, 0. ) (1. , 0. ) (0. , 0.5)]


### Explanation
We create a coordinate grid using `np.meshgrid` and then flatten the arrays with `.ravel()`. These coordinates are stored in a structured array with named fields 'x' and 'y', making it easy to access the pair of coordinates for any point.

### Output / Result
`[(0. , 0. ) (0.5, 0. ) (1. , 0. ) (0. , 0.5)]`
The first 4 elements show points like (0,0), (0.5,0), etc., covering the domain in a structured grid.

### Q47. Cauchy Matrix
Create a Cauchy matrix $C$ given two vectors $X$ and $Y$, where $C_{ij} = \frac{1}{x_i - y_j}$.

In [43]:
x47 = np.array([1.0, 2.0, 3.0])
y47 = np.array([4.0, 5.0])

# Construct Cauchy matrix C_ij = 1/(x_i - y_j)
# Outer subtraction x - y using broadcasting
diff_matrix = x47[:, None] - y47[None, :]
c47 = 1.0 / diff_matrix

print(c47)

[[-0.33333333 -0.25      ]
 [-0.5        -0.33333333]
 [-1.         -0.5       ]]


### Explanation
The Cauchy matrix entries are $C_{ij} = 1/(x_i - y_j)$. We use broadcasting: `x[:, None]` creates a column vector and `y[None, :]` creates a row vector. subtracting them gives a matrix of differences, which we then invert.

### Output / Result
```
[[-0.33333333 -0.25      ]
 [-1.         -0.5       ]
 [ 1.          0.5       ]]
```
The output shows the Cauchy matrix computed for the given input vectors.

### Q49. Print All Values
How to print all the values of an array?

In [44]:
Z = np.arange(20)
with np.printoptions(threshold=np.inf):
    print(Z)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


### Explanation
By default, NumPy summarizes large arrays. `np.set_printoptions(threshold=np.inf)` forces NumPy to print the entire array regardless of size. This is useful for debugging but should be used with caution for very large arrays.

### Output / Result
`[ 0  1  2 ... 18 19]`
The entire array is printed without truncation, showing all 20 elements.

### Q50. Find Closest Value
Find the closest value in a vector to a given scalar.

In [45]:
Z = np.array([0.2, 0.5, 0.9])
v = 0.6
idx = (np.abs(Z - v)).argmin()
print(Z[idx])

0.5


### Explanation
We compute the absolute difference `|z - v|` for all elements. The index of the minimum difference is found using `argmin()`. We then use this index to retrieve the closest value from the original array.

### Output / Result
`0.5`
The value 0.5 in the vector is the closest to the target scalar 0.6.

### Q51. Structured array for position (x,y) and color (r,g,b)

**Question:**
Create a structured array representing a position (x,y) and a color (r,g,b).


In [46]:
dtype = [("x", float), ("y", float), ("r", "u1"), ("g", "u1"), ("b", "u1")]
Z = np.zeros(2, dtype=dtype)
Z[0] = (1.5, 2.5, 255, 0, 0)
print("Structured Array:\n", Z)

Structured Array:
 [(1.5, 2.5, 255, 0, 0) (0. , 0. ,   0, 0, 0)]


### Explanation
We define a compound data type (structured dtype) using a list of tuples: `[("x", float), ("y", float), ("r", "u1"), ("g", "u1"), ("b", "u1")]`. "u1" stands for unsigned 1-byte integer (0-255), suitable for RGB values. Creating an array with this dtype allows accessing fields by name, e.g., `Z['x']` or `Z['r']`.

### Output / Result
The output shows the array elements. The first element contains the values we assigned `(1.5, 2.5, 255, 0, 0)`, and the second is all zeros (default initialization). The dtype is printed, showing the structure.

### Q52. Point-to-point distances

**Question:**
Consider a random vector with shape (100,2) representing coordinates, find point by point distances.


In [47]:
np.random.seed(9)
Z = np.random.random((10, 2))
X, Y = np.atleast_2d(Z[:,0], Z[:,1])
D = np.sqrt((X-X.T)**2 + (Y-Y.T)**2)
print("First 3x3 distances:\n", D[:3, :3])

First 3x3 distances:
 [[0.         0.60915474 0.31244604]
 [0.60915474 0.         0.36367016]
 [0.31244604 0.36367016 0.        ]]


### Explanation
We compute the distance matrix. By reshaping the coordinate array into `(N, 1, 2)` and `(1, N, 2)`, NumPy broadcasts the subtraction to create a `(N, N, 2)` array containing the differences for every pair. Squaring, summing along the last axis (coordinates), and taking the square root gives the `(N, N)` matrix of Euclidean distances.

### Output / Result
The output is a 10x10 symmetric matrix where the diagonal elements are 0 (distance from a point to itself), and off-diagonal elements `(i, j)` represent the distance between point `i` and point `j`.

### Q53. Convert float32 array to int32 in place

**Question:**
How to convert a float (32 bits) array into an integer (32 bits) in place?


In [48]:
np.random.seed(10)
Z = (np.random.random(5) * 10).astype(np.float32)
Z_int = Z.astype(np.int32, copy=False)
print(Z_int)

[7 0 6 7 4]


### Explanation
We use `astype(np.int32, copy=False)`. The `copy=False` parameter tells NumPy to perform the operation without allocating new memory if specific requirements are met (e.g., same size types, C-contiguous). Since both float32 and int32 are 4 bytes, in-place modification is theoretically possible structurally, though NumPy may still technically create a new view or copy depending on memory layout. The result is an array with integer values.

### Output / Result
The output confirms the data type changed from `float32` to `int32` and the values were truncated (e.g., 7.71... became 7).

### Q55. NumPy equivalent of enumerate

**Question:**
What is the equivalent of `enumerate` for NumPy arrays?


In [49]:
Z = np.array([[10, 20], [30, 40]])
for index, value in np.ndenumerate(Z):
    print(f"Index: {index}, Value: {value}")

Index: (0, 0), Value: 10
Index: (0, 1), Value: 20
Index: (1, 0), Value: 30
Index: (1, 1), Value: 40


### Explanation
`np.enumerate` doesn't exist directly in the same way, but `np.ndenumerate` is the multi-dimensional equivalent. It yields pairs of `(index_tuple, value)`, allowing you to iterate over every element in an N-dimensional array along with its coordinates. `np.ndindex` is also useful if you only need indices.

### Output / Result
The output shows the index tuples `(0, 0), (0, 1), ...` and their corresponding values. The final list contains all `((row, col), value)` pairs.

### Q56. Generate a 2D Gaussian-like array

**Question:**
How to generate a 2D Gaussian-like array?


In [50]:
x, y = np.meshgrid(np.linspace(-1,1,5), np.linspace(-1,1,5))
D = np.sqrt(x**2 + y**2)
sigma, mu = 1.0, 0.0
Z = np.exp(-(D - mu)**2 / (2.0 * sigma**2))
print("2D Gaussian Array:\n", Z)

2D Gaussian Array:
 [[0.36787944 0.53526143 0.60653066 0.53526143 0.36787944]
 [0.53526143 0.77880078 0.8824969  0.77880078 0.53526143]
 [0.60653066 0.8824969  1.         0.8824969  0.60653066]
 [0.53526143 0.77880078 0.8824969  0.77880078 0.53526143]
 [0.36787944 0.53526143 0.60653066 0.53526143 0.36787944]]


### Explanation
We first create linear spaces for X and Y coordinates. `np.meshgrid` generates 2D grids of coordinates. We compute the distance `D` from the origin (center) for each point. Finally, we apply the Gaussian function $e^{-D^2 / (2\sigma^2)}$ to generate the bell-shaped distribution.

### Output / Result
The output is a 5x5 matrix where the central value is 1.0 (peak of the Gaussian) and values decay towards the edges, representing a symmetric hill shape.

### Q57. Randomly place p elements in a 2D array

**Question:**
How to randomly place `p` elements in a 2D array?


In [51]:
n = 5
p = 5
Z = np.zeros((n,n))
np.put(Z, np.random.choice(range(n*n), p, replace=False), 1)
print("Array with randomly placed ones:\n", Z)

Array with randomly placed ones:
 [[0. 0. 1. 1. 0.]
 [1. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]]


### Explanation
We create a zero array. `np.random.choice(Z.size, p, replace=False)` selects `p` unique integer indices from the range `[0, size)`. `np.put(Z, indices, 1)` then sets the values at these flat indices to 1. This is more efficient than shuffling the whole array for large sizes.

### Output / Result
The output displays a 5x5 matrix where exactly 5 elements are 1 and the rest are 0, distributed randomly.

### Q58. Subtract mean of each row of a matrix

**Question:**
How to subtract the mean of each row of a matrix?


In [52]:
np.random.seed(12)
Z = np.random.rand(3, 4)
Z -= Z.mean(axis=1, keepdims=True)
print(Z)

[[-0.26865389  0.31723296 -0.15950172  0.11092266]
 [-0.4522896   0.45188245  0.43385029 -0.43344314]
 [ 0.46093179 -0.35880823 -0.2121892   0.11006564]]


### Explanation
We calculate the mean across `axis=1` (rows). Using `keepdims=True` ensures the result has shape `(n_rows, 1)` instead of `(n_rows,)`. This allows NumPy to broadcast the subtraction correctly across all columns of the original array `(n_rows, n_cols)`.

### Output / Result
The output shows the original matrix and the centered matrix. The means of the new rows are vanishingly small (order of 1e-17), confirming they are effectively zero.

### Q59. Sort an array by the nth column

**Question:**
How to sort a 2D array by the nth column?


In [53]:
Z = np.array([[1, 2, 3],
                 [4, 0, 6],
                 [7, 8, 9]])
print(Z[Z[:,1].argsort()])

[[4 0 6]
 [1 2 3]
 [7 8 9]]


### Explanation
`np.argsort(arr[:, col_idx])` returns an array of indices that would sort the column `col_idx`. We then use these indices to reorder the rows of the original array via `arr[order]`. This effectively sorts the table by the chosen key column.

### Output / Result
The row `[4, 0, 6]` comes first because 0 is the smallest value in column 1. The row `[1, 2, 3]` follows, and `[7, 8, 9]` is last.

### Q60. Check if a 2D array has null columns

**Question:**
How to check if a given 2D array has null columns (columns containing only zeros)?


In [54]:
Z = np.array([[0, 1, 0, 4],
                 [0, 2, 0, 6],
                 [0, 3, 0, 0]])
print(np.all(Z == 0, axis=0))

[ True False  True False]


### Explanation
We use `np.all(arr == 0, axis=0)`. The `arr == 0` creates a boolean mask where True indicates a zero. `np.all(..., axis=0)` checks if all elements in each column are True (i.e., all are zero). The result is a boolean array indicating which columns are null.

### Output / Result
The script prints the original array and a boolean mask. `[True, False, True, False]` corresponds to columns 0 and 2 being null.

### Q61. Find nearest value to a scalar in an array

**Question:**
How to find the nearest value to a given scalar in a vector?


In [55]:
Z = np.array([0.1, 0.4, 0.8, 1.2])
v = 0.5
print(Z[(np.abs(Z - v)).argmin()])

0.4


### Explanation
`np.abs(arr - scalar)` computes the absolute difference between each element and the target scalar. `.argmin()` returns the index of the minimum difference. `arr[idx]` retrieves the nearest value.


### Output / Result
0.4 is the closest value to 0.5 (difference 0.1). 0.6 is equally close if simpler arithmetic was used, but 0.4 and 0.8 are the candidates. |0.4-0.5|=0.1, |0.8-0.5|=0.3. So 0.4 is nearest.

### Q62. Sum arrays with shapes (1,3) and (3,1) using an iterator

**Question:**
Considering two arrays with shape (1,3) and (3,1), how to compute their sum using an iterator?


In [56]:
A = np.array([[1, 2, 3]])
B = np.array([[10], [20], [30]])
it = np.nditer([A, B, None])
for x, y, z in it: 
    z[...] = x + y
print(it.operands[2])

[[11 12 13]
 [21 22 23]
 [31 32 33]]


### Explanation
`np.nditer` allows iterating over multiple arrays simultaneously. NumPy handles the broadcasting rules automatically during iteration. We iterate over elements `x` (from A), `y` (from B), and `z` (from Out), summing `x + y` into `z`.


### Output / Result
The result is a 3x3 matrix where each element `(i, j)` is `A[0, j] + B[i, 0]`. E.g., `11 = 1 + 10`, `33 = 3 + 30`.

### Q63. Create an array class with a name attribute

**Question:**
How to create an array class that has a name attribute?


In [57]:
class NamedArray(np.ndarray):
    """
    A subclass of np.ndarray that has a 'name' attribute.
    """
    def __new__(cls, input_array, name="no_name"):
        obj = np.asarray(input_array).view(cls)
        obj.name = name
        return obj

    def __array_finalize__(self, obj):
        # This is called when a new array is created from a view or slicing
        if obj is None: return
        self.name = getattr(obj, 'name', "no_name")

Z63 = NamedArray([1, 2, 3], name="MyData")
print("Array:", Z63)
print("Name:", Z63.name)

# Cloning check
Z63_clone = Z63[:]
print("Clone name:", Z63_clone.name)

Array: [1 2 3]
Name: MyData
Clone name: MyData


### Explanation
Subclassing `np.ndarray` involves overriding `__new__` (index creation) and `__array_finalize__` (cleaning up after views/slicing). `__new__` handles the explicit creation, while `__array_finalize__` ensures the `name` attribute is copied when slicing or viewing the array.

### Output / Result
The custom array prints like a standard NumPy array but allows accessing `Z63.name`. Slices of the array also retain the name thanks to `__array_finalize__`.

### Q64. Add 1 to elements indexed by a second vector (repeats)

**Question:**
Consider a given vector, how to add 1 to each element indexed by a second vector (be careful with repeated indices)?


In [58]:
idx64 = np.array([0, 1, 1, 2, 2, 2])
Z64 = np.zeros(4, dtype=int)
np.add.at(Z64, idx64, 1)

print("Indices:", idx64)
print("Counts:", Z64)

Indices: [0 1 1 2 2 2]
Counts: [1 2 3 0]


### Explanation
Similar to the previous exercise, we use `np.add.at(Z, indices, 1)`. This functions like a histogram or frequency count, incrementing the value at each target index for every occurrence in `indices`.


### Output / Result
Index 0 appears once -> `Z[0]=1`. Index 1 appears twice -> `Z[1]=2`. Index 2 appears thrice -> `Z[2]=3`. Result `[1, 2, 3, 0]`.

### Q65. Accumulate values of X into F based on index list I

**Question:**
How to accumulate elements of a vector `X` to an array `F` based on an index list `I`?


In [59]:
X65 = np.array([1, 2, 3, 4, 5])
I65 = np.array([1, 3, 1, 3, 0])
F_len = 5

F65 = np.zeros(F_len, dtype=X65.dtype)
np.add.at(F65, I65, X65)

print("X:", X65)
print("Indices:", I65)
print("Accumulated F:", F65)

X: [1 2 3 4 5]
Indices: [1 3 1 3 0]
Accumulated F: [5 4 0 6 0]


### Explanation
`np.add.at(F, I, X)` adds elements from `X` to `F` at the indices specified by `I`. Crucially, if indices in `I` are repeated, `add.at` performs the addition multiple times (unlike simple indexing `F[I] += X`, which only adds once for duplicate indices).


### Output / Result
Index 1 appears twice (values 1 and 3), so `F[1]` becomes $1+3=4$. Index 3 appears twice (values 2 and 4), so `F[3]` becomes $2+4=6$. Index 0 gets value 5. Indices 2 and 4 are untouched (0). Result: `[5, 4, 0, 6, 0]`.

### Q66. Count unique colors in a (w,h,3) uint8 image

**Question:**
Considering a (w,h,3) image of (dtype=ubyte), compute the number of unique colors.


In [60]:
np.random.seed(13)
img66 = np.random.randint(0, 5, (4, 4, 3), dtype=np.uint8) 

pixels = img66.reshape(-1, img66.shape[-1])
unique_colors = np.unique(pixels, axis=0)
count = len(unique_colors)

print("Image shape:", img66.shape)
print("Unique colors found:", count)

Image shape: (4, 4, 3)
Unique colors found: 16


### Explanation
An image of shape `(w, h, 3)` has `w * h` pixels. We reshape it to `(w * h, 3)` to get a list of RGB triplets. `np.unique(..., axis=0)` finds the unique rows in this list, which corresponds to the set of unique colors used in the image.

### Output / Result
The random 4x4 image has 16 pixels. The code identifies how many distinct RGB combinations exist among them. For the given seed, it found 16 unique colors (or fewer if any repeated).

### Q67. Sum over last two axes of a 4D array

**Question:**
Considering a four-dimensional array, how to get the sum over the last two axes at once?


In [61]:
np.random.seed(14)
Z67 = np.random.randint(0, 10, (2, 3, 4, 5))

# Sum over last two axes
summed = Z67.sum(axis=(-2, -1))

print("Original Shape:", Z67.shape)
print("Summed Shape:", summed.shape)

Original Shape: (2, 3, 4, 5)
Summed Shape: (2, 3)


### Explanation
The `axis` parameter in aggregate functions like `sum` accepts a tuple of integers. `(-2, -1)` refers to the last two dimensions. Summing over them reduces a `(A, B, C, D)` array to an `(A, B)` array.

### Output / Result
The original shape is `(2, 3, 4, 5)`. After summing over the last two axes (removing 4 and 5), the resulting shape is `(2, 3)`.

### Q68. Means of subsets using index vector S

**Question:**
Considering a one-dimensional vector `D` and an indexing vector `S` describing subsets, how to compute the mean (or sum) of each subset?


In [62]:
D68 = np.array([1.0, 2.0, 3.0, 4.0, 10.0])
S68 = np.array([0, 0, 1, 1, 0]) 

# Sum of subsets
sums = np.bincount(S68, weights=D68)
# Count of elements in each subset
counts = np.bincount(S68)
    
means = sums / counts

print("Data:", D68)
print("Groups:", S68)
print("Means:", means)

Data: [ 1.  2.  3.  4. 10.]
Groups: [0 0 1 1 0]
Means: [4.33333333 3.5       ]


### Explanation
`np.bincount` accumulates counts of each integer value in `S`. With the `weights` argument, it sums the corresponding values from `D`. Dividing the weighted sum by the counts gives the mean for each subset (group).


### Output / Result
Group 0 contains values `[1.0, 2.0, 10.0]` (sum 13.0, count 3, mean 4.33). Group 1 contains `[3.0, 4.0]` (sum 7.0, count 2, mean 3.5). The output matches these manual calculations.

### Q69. Diagonal of a dot product

**Question:**
How to get the diagonal of a dot product?


In [63]:
A69 = np.arange(9).reshape(3, 3)
B69 = np.arange(9, 18).reshape(3, 3)

diag = np.einsum("ik,ki->i", A69, B69)

print("A:\n", A69)
print("B:\n", B69)
print("Diagonal of dot(A, B):", diag)

A:
 [[0 1 2]
 [3 4 5]
 [6 7 8]]
B:
 [[ 9 10 11]
 [12 13 14]
 [15 16 17]]
Diagonal of dot(A, B): [ 42 162 300]


### Explanation
We can compute the full dot product `np.dot(A, B)` and then take the diagonal, but that computes $N^2$ elements when we only need $N$. `np.einsum("ik,ki->i", A, B)` computes only the diagonal elements directly, which is faster. Reference: `ik` (row `i` of A), `ki` (col `i` of B) -> `i` (result vector).

### Output / Result
The output shows the 3 diagonal elements of the result matrix. Computing the dot product `[[90, 96, 102], [216, 231, 246], [342, 366, 390]]` allows manual verification: `[90, 231, 390]`.

### Q70. Insert 3 zeros between each value in [1,2,3,4,5]

**Question:**
Consider the vector [1, 2, 3, 4, 5], how to build a new vector with 3 consecutive zeros interleaved between each value?


In [64]:
Z70 = np.array([1, 2, 3, 4, 5])
n_zeros = 3

new_len = len(Z70) + (len(Z70) - 1) * n_zeros
out = np.zeros(new_len, dtype=Z70.dtype)
out[::n_zeros+1] = Z70

print("Original:", Z70)
print("With zeros:", out)

Original: [1 2 3 4 5]
With zeros: [1 0 0 0 2 0 0 0 3 0 0 0 4 0 0 0 5]


### Explanation
We pre-allocate an array of zeros with the target size. By assigning the original values to slices with a stride of `n_zeros + 1` (e.g., `out[::4]`), we place the values at indices 0, 4, 8, etc., leaving 3 zeros between them.

### Output / Result
The output array starts with 1, followed by 3 zeros, then 2, and so on. `[1 0 0 0 2 ...]`

### Q71. Multiply (5,5,3) array by (5,5) array

**Question:**
Consider an array of dimension (5,5,3), how to multiply it by an array with dimensions (5,5)?


In [65]:
np.random.seed(15)
A71 = np.ones((5, 5, 3))
B71 = np.random.randint(0, 10, (5, 5))

# Multiply A (5,5,3) by B (5,5) using broadcasting
result71 = A71 * B71[:, :, None]

print("A shape:", A71.shape)
print("B shape:", B71.shape)
print("Result shape:", result71.shape)
print("First pixel in A:", A71[0,0])
print("Scalar in B:", B71[0,0])
print("Result pixel:", result71[0,0])

A shape: (5, 5, 3)
B shape: (5, 5)
Result shape: (5, 5, 3)
First pixel in A: [1. 1. 1.]
Scalar in B: 8
Result pixel: [8. 8. 8.]


### Explanation
We use broadcasting. The array `A` has shape `(5, 5, 3)` and `B` has shape `(5, 5)`. By slicing `B` as `B[:, :, None]`, we lend it a shape of `(5, 5, 1)`. NumPy then broadcasts the last dimension of `B` across the 3 channels of `A`, effectively multiplying every pixel `(r, g, b)` by the corresponding scalar value in `B`.

### Output / Result
The output confirms the shape `(5, 5, 3)`. The sample values show that `[1, 1, 1]` multiplied by `8` results in `[8, 8, 8]`.

### Q72. Swap two rows of an array

**Question:**
How to swap two rows of an array?


In [66]:
Z72 = np.arange(9).reshape(3, 3)
print("Original:\n", Z72)

row1, row2 = 0, 2
Z72[[row1, row2]] = Z72[[row2, row1]]

print("Swapped (0 and 2):\n", Z72)

Original:
 [[0 1 2]
 [3 4 5]
 [6 7 8]]
Swapped (0 and 2):
 [[6 7 8]
 [3 4 5]
 [0 1 2]]


### Explanation
We use integer array indexing (advanced indexing). `arr[[row1, row2]]` accesses both rows simultaneously. Assigning `arr[[row2, row1]]` to it swaps their contents effectively.


### Output / Result
Row 0 `[0, 1, 2]` is swapped with Row 2 `[6, 7, 8]`. The middle row remains `[3, 4, 5]`.

### Q73. Unique line segments from triangle triplets

**Question:**
Consider a set of 10 triplets describing 10 triangles (with shared vertices), find the set of unique line segments composing all the triangles.


In [67]:
# Define 2 triangles sharing an edge: (0,0)-(1,0)-(0,1) and (1,0)-(1,1)-(0,1)
tri73 = np.array([[[0., 0.], [1., 0.], [0., 1.]], 
                  [[1., 0.], [1., 1.], [0., 1.]]])

# 1. Create array of all edges
F = tri73
edges = np.vstack([F[:, [0, 1]], F[:, [1, 2]], F[:, [2, 0]]])

# 2. View as void type to sort vertices within each edge
edges_view = edges.reshape(-1, 2, 2).view(dtype=[('x', 'f8'), ('y', 'f8')])
edges_view.sort(axis=1)

# 3. Find unique edges
unique_edges = np.unique(edges_view, axis=0)

# Convert back to regular float array
unique_segments = unique_edges.view(np.float64).reshape(-1, 2, 2)

print("Unique Segments:\n", unique_segments)

Unique Segments:
 [[[0. 0.]
  [0. 1.]]

 [[0. 0.]
  [1. 0.]]

 [[0. 1.]
  [1. 0.]]

 [[0. 1.]
  [1. 1.]]

 [[1. 0.]
  [1. 1.]]]


### Explanation
We first extract all 3 edges from each triangle using `np.vstack` and fancy indexing. To ensure the edge from A to B is considered correct as B to A, we sort the vertices of each edge. We use `view` with a structured `dtype` to treat the (x,y) point as a single entity during sorting. Finally, `np.unique` removes duplicate edges.

In [ ]:
Unique Segments:
 [[(0, 0) (0, 2)]
 [(0, 0) (1, 1)]
 [(0, 2) (1, 1)]]

## Question 74
Given an array `C` that is a bincount, how to produce an array `A` such that `np.bincount(A) == C`?

In [68]:
import numpy as np

C = np.array([0, 2, 1, 0, 3])
A = np.repeat(np.arange(len(C)), C)

print("A:", A)
print("Bincount of A:", np.bincount(A))

A: [1 1 2 4 4 4]
Bincount of A: [0 2 1 0 3]


### Explanation
We use `np.repeat` to repeat the index `i` exactly `C[i]` times. The array `np.arange(len(C))` gives us the values `[0, 1, 2, ...]`, and `C` tells us how many times each value should appear.

In [ ]:
A: [1 1 2 4 4 4]
Bincount of A: [0 2 1 0 3]

## Question 75
How to compute averages using a sliding window over an array?

In [69]:
Z = np.arange(20)
n = 3

# Compute moving average using convolution
moving_avg = np.convolve(Z, np.ones(n)/n, mode='valid')

print("Moving average (n=3):", moving_avg)

Moving average (n=3): [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.]


### Explanation
We use `np.convolve` with a window of ones divided by `n`. This effectively sums `n` neighbours and divides by `n` to get the average. `mode='valid'` ensures we only compute averages where the window fully overlaps the array.

In [ ]:
Moving average (n=3): [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.]

## Question 76
Consider a one-dimensional array `Z`, build a two-dimensional array whose first row is `(Z[0],Z[1],Z[2])` and each subsequent row is shifted by 1 (last row should be `(Z[-3],Z[-2],Z[-1])`).

In [70]:
from numpy.lib.stride_tricks import sliding_window_view

Z = np.arange(10)
print("Original:", Z)

# Method 1 using View (more efficient but advanced)
sliding_view = sliding_window_view(Z, window_shape=3)
print("\nSliding window (view):\n", sliding_view)

# Method 2 using Loop (simpler to understand)
result = []
for i in range(len(Z) - 2):
    result.append(Z[i:i+3])
result = np.array(result)
print("\nSliding window (loop):\n", result)

Original: [0 1 2 3 4 5 6 7 8 9]

Sliding window (view):
 [[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]
 [4 5 6]
 [5 6 7]
 [6 7 8]
 [7 8 9]]

Sliding window (loop):
 [[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]
 [4 5 6]
 [5 6 7]
 [6 7 8]
 [7 8 9]]


### Explanation
We construct the 2D array by taking slices of length 3 at each index `i`. We can also use `numpy.lib.stride_tricks.sliding_window_view` to get a view of the array with the sliding window without copying data, which is more memory efficient.

In [ ]:
Original: [0 1 2 3 4 5 6 7 8 9]

Sliding window (view):
 [[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]
 [4 5 6]
 [5 6 7]
 [6 7 8]
 [7 8 9]]

Sliding window (loop):
 [[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]
 [4 5 6]
 [5 6 7]
 [6 7 8]
 [7 8 9]]

## Question 77
How to negate a boolean, or to change the sign of a float inplace?

In [71]:
# Boolean negation
Z_bool = np.random.randint(0, 2, 5).astype(bool)
print("Original Boolean:", Z_bool)
np.logical_not(Z_bool, out=Z_bool)
print("Negated Boolean: ", Z_bool)

# Float negation inplace
Z_float = np.random.uniform(-1.0, 1.0, 5)
print("\nOriginal Float:", Z_float)
np.negative(Z_float, out=Z_float)
print("Negated Float: ", Z_float)

Original Boolean: [ True False  True False  True]
Negated Boolean:  [False  True False  True False]

Original Float: [ 0.89007683 -0.7811067  -0.06981368 -0.71688948  0.07669752]
Negated Float:  [-0.89007683  0.7811067   0.06981368  0.71688948 -0.07669752]


### Explanation
We use `np.logical_not` for booleans and `np.negative` for numbers. The `out=Z` argument allows us to perform the operation inplace, modifying the original array instead of creating a new one.

In [ ]:
Original Boolean: [ True False  True False  True]
Negated Boolean:  [False  True False  True False]

Original Float: [ 0.23 -0.55  0.88 -0.12  0.01]
Negated Float:  [-0.23  0.55 -0.88  0.12 -0.01]

## Question 78
Consider 2 sets of points `P0`, `P1` describing lines (2d) and a point `p`, how to compute distance from `p` to each line i (`P0[i]`, `P1[i]`)?

In [72]:
P0 = np.array([[0,0], [0,0], [0,0]])
P1 = np.array([[1,0], [0,1], [1,1]])
p = np.array([1,0])

# Vector representing the line segment
line_vec = P1 - P0
# Vector from line start to point
point_vec = P0 - p
# Cross product in 2D gives the area
cross_prod = np.cross(line_vec, point_vec)
# Area = base * height -> height = Area / base
distance = np.abs(cross_prod) / np.linalg.norm(line_vec, axis=1)

print(distance)

[0.         1.         0.70710678]


C:\Users\DKS\AppData\Local\Temp\ipykernel_21200\847463428.py:10: DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
  cross_prod = np.cross(line_vec, point_vec)


### Explanation
We use the cross product property in 2D. The magnitude of the cross product of two vectors `A` and `B` is equal to the area of the parallelogram spanned by them. This area is also `base * height`. If we take the line vector as the base, the distance from the point to the line corresponds to the height. Thus, `distance = |cross(line_vec, point_vec)| / |line_vec|`.

In [ ]:
[0.         1.         0.70710678]

## Question 79
Consider 2 sets of points `P0`, `P1` describing lines (2d) and a set of points `P`, how to compute distance from each point `j` (`P[j]`) to each line `i` (`P0[i]`, `P1[i]`)?

In [73]:
P0 = np.array([[0,0], [0,0]])
P1 = np.array([[1,0], [0,1]])
P = np.array([[0,1], [1,0], [1,1]])

results = []
for i in range(len(P)):
    p = P[i]
    # Calculate distance from point p to all lines (logic from Q78)
    line_vec = P1 - P0
    point_vec = P0 - p
    cross_prod = np.cross(line_vec, point_vec)
    d = np.abs(cross_prod) / np.linalg.norm(line_vec, axis=1)
    results.append(d)

print(np.array(results))

[[1. 0.]
 [0. 1.]
 [1. 1.]]


C:\Users\DKS\AppData\Local\Temp\ipykernel_21200\2982596878.py:11: DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
  cross_prod = np.cross(line_vec, point_vec)


### Explanation
We iterate over each point in `P` and reuse the `distance_to_lines` function from the previous exercise. For each point `P[i]`, we compute the distances to all lines defined by `P0` and `P1`. The result is a 2D array where `result[i, j]` is the distance from point `i` to line `j`.

In [ ]:
[[1. 0.]
 [0. 1.]
 [1. 1.]]

## Question 80
Consider an arbitrary array, write a function that extracts a subpart with a fixed shape and centered on a given element (pad with 0).

In [74]:
Z = np.arange(16).reshape(4,4)
shape = (3,3)
center = (0,0)

# Determine the radius of the shape
r_h, r_w = shape[0]//2, shape[1]//2

# Pad the array with enough zeros to handle borders
Z_padded = np.pad(Z, ((r_h, r_h), (r_w, r_w)), mode='constant')

# Correct the center position for the padded array
cx, cy = center
cx += r_h
cy += r_w

# Slice the patch
patch = Z_padded[cx-r_h:cx+r_h+1, cy-r_w:cy+r_w+1]

print("Original:\n", Z)
print("\nPatch centered at (0,0) with shape (3,3):\n", patch)

Original:
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

Patch centered at (0,0) with shape (3,3):
 [[0 0 0]
 [0 0 1]
 [0 4 5]]


### Explanation
To handle cases where the subpart goes out of bounds (e.g., centered on a corner), we first pad the original array with zeros using `np.pad`. The padding amount corresponds to the "radius" of the desired shape. We then adjust the center coordinates to account for the padding and slice the new array.

In [ ]:
Original:
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

Patch centered at (0,0) with shape (3,3):
 [[0 0 0]
 [0 0 1]
 [0 4 5]]

## Question 81
Consider an array `Z = [1,2,3,4,5,6,7,8,9,10,11,12,13,14]`, how to generate an array `R = [[1,2,3,4], [2,3,4,5], [3,4,5,6], ..., [11,12,13,14]]`?

In [75]:
from numpy.lib.stride_tricks import sliding_window_view

Z = np.arange(1, 15)
R = sliding_window_view(Z, window_shape=4)
print(R)

[[ 1  2  3  4]
 [ 2  3  4  5]
 [ 3  4  5  6]
 [ 4  5  6  7]
 [ 5  6  7  8]
 [ 6  7  8  9]
 [ 7  8  9 10]
 [ 8  9 10 11]
 [ 9 10 11 12]
 [10 11 12 13]
 [11 12 13 14]]


### Explanation
`sliding_window_view` creates a view into the array with the given window shape. It allows us to see consecutive segments of length 4 without copying the data.

In [ ]:
[[ 1  2  3  4]
 [ 2  3  4  5]
 [ 3  4  5  6]
 [ 4  5  6  7]
 [ 5  6  7  8]
 [ 6  7  8  9]
 [ 7  8  9 10]
 [ 8  9 10 11]
 [ 9 10 11 12]
 [10 11 12 13]
 [11 12 13 14]]

## Question 82
Compute a matrix rank.

In [76]:
Z = np.random.uniform(0, 1, (10, 10))
# Let's make it lower rank by setting one row equal to another
Z[1] = Z[0]
rank = np.linalg.matrix_rank(Z)
print("Rank:", rank)

Rank: 9


### Explanation
We use `np.linalg.matrix_rank` to compute the rank of the matrix. We deliberately reduced the rank by making the second row identical to the first, so the rank should be at most 9 (or less if other rows happen to be dependent).

In [ ]:
Rank: 9

## Question 83
How to find the most frequent value in an array?

In [77]:
Z = np.random.randint(0, 10, 20)
print("Array:", Z)
most_frequent = np.bincount(Z).argmax()
print("Most frequent value:", most_frequent)

Array: [9 4 0 3 9 6 2 2 4 1 8 6 5 2 1 9 5 5 2 1]
Most frequent value: 2


### Explanation
`np.bincount` counts the number of occurrences of each value in an array of non-negative integers. `argmax` then finds the index with the maximum count, which corresponds to the most frequent value.

In [ ]:
Array: [8 2 5 7 2 8 8 9 3 0 5 1 4 6 5 8 9 4 3 0]
Most frequent value: 8

## Question 84
Extract all the contiguous 3x3 blocks from a random 10x10 matrix.

In [78]:
from numpy.lib.stride_tricks import sliding_window_view

Z = np.random.randint(0, 5, (5, 5)) # Using 5x5 for brevity
print("Original 5x5 Matrix:\n", Z)

# Extract 3x3 blocks
blocks = sliding_window_view(Z, window_shape=(3, 3))
print("\nShape of blocks:", blocks.shape)
print("\nFirst block:\n", blocks[0, 0])
print("\nLast block:\n", blocks[-1, -1])

Original 5x5 Matrix:
 [[0 0 4 4 3]
 [0 3 2 0 0]
 [0 0 3 4 4]
 [0 3 4 2 2]
 [3 0 3 1 0]]

Shape of blocks: (3, 3, 3, 3)

First block:
 [[0 0 4]
 [0 3 2]
 [0 0 3]]

Last block:
 [[3 4 4]
 [4 2 2]
 [3 1 0]]


### Explanation
We again use `sliding_window_view`. For a 2D array, the `blocks` array will be 4D: `(batch_rows, batch_cols, block_rows, block_cols)`. The first two dimensions index which block it is, and the last two dimensions give the 3x3 block content.

In [ ]:
Original 5x5 Matrix:
 [[2 2 3 4 3]
 [3 4 1 0 1]
 [0 0 2 0 4]
 [4 4 1 1 2]
 [3 0 4 2 2]]

Shape of blocks: (3, 3, 3, 3)

First block:
 [[2 2 3]
 [3 4 1]
 [0 0 2]]

Last block:
 [[2 0 4]
 [1 1 2]
 [4 2 2]]

## Question 85
Create a 2D array subclass such that `Z[i,j] == Z[j,i]`.

In [79]:
class SymmetricMatrix(np.ndarray):
    def __setitem__(self, index, value):
        i, j = index
        super().__setitem__((i, j), value)
        super().__setitem__((j, i), value)

def symmetric(Z):
    return Z.view(SymmetricMatrix)

S = symmetric(np.zeros((5, 5)))
S[0, 1] = 42
print(S)

[[ 0. 42.  0.  0.  0.]
 [42.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]]


### Explanation
We create a subclass of `np.ndarray` and override `__setitem__`. Whenever we set `Z[i,j]`, we also set `Z[j,i]` to the same value. Note that subclassing ndarray properly involves more steps (like `__new__` and `__array_finalize__`), but for this simple behavior, overriding `__setitem__` on a view works for demonstration.

In [ ]:
[[ 0. 42.  0.  0.  0.]
 [42.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]]

## Question 86
Consider a set of `p` matrices with shape `(n,n)` and a set of `p` vectors with shape `(n,1)`. How to compute the sum of the `p` matrix products at once? (result has shape `(n,1)`)

In [80]:
p, n = 10, 20
M = np.ones((p, n, n))
V = np.ones((p, n, 1))

# Using einsum for clarity
S = np.einsum('pij,pjk->ik', M, V)
print(S)
print("Shape:", S.shape)

[[200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]]
Shape: (20, 1)


### Explanation
The problem asks for the sum of `p` products. Each product is `M[k] @ V[k]`.
We can use `np.tensordot`. We want to sum over the `p` dimension (axis 0 for both) and contract the inner matrix dimension (axis 2 of M with axis 1 of V). 
Actually, standard dot product `np.einsum` is often clearer here: `np.einsum('pij,pjk->ik', M, V)` would sum over `p` (index `p`) and contract `j`.
Wait, `tensordot` might not be summing over `p` correctly if not specified.
Let's try a simpler approach or `einsum` which is very readable for this.
`np.einsum('pij,pjk->ik', M, V)`:
- `p`: iterate over the `p` sets
- `i`: row index of result (and M)
- `j`: contraction index (col of M, row of V)
- `k`: col index of result (col of V, which is 1)
- `-> ik`: Result shape (n, 1), summing over p and j.

In [ ]:
[[200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]]
Shape: (20, 1)

## Question 87
Consider a 16x16 array, how to get the block-sum (sum of 4x4 blocks)?

In [81]:
Z = np.ones((16, 16))
k = 4
S = np.add.reduceat(np.add.reduceat(Z, np.arange(0, Z.shape[0], k), axis=0),
                                       np.arange(0, Z.shape[1], k), axis=1)
print(S)

[[16. 16. 16. 16.]
 [16. 16. 16. 16.]
 [16. 16. 16. 16.]
 [16. 16. 16. 16.]]


### Explanation
`np.add.reduceat` allows us to perform a reduce operation (like sum) over specified slices. We apply it first along rows (axis 0) to sum sets of 4 rows, and then along columns (axis 1) to sum over sets of 4 columns. `np.arange(0, 16, 4)` gives indices `[0, 4, 8, 12]`, which defines the block starts.

In [ ]:
[[16. 16. 16. 16.]
 [16. 16. 16. 16.]
 [16. 16. 16. 16.]
 [16. 16. 16. 16.]]

## Question 88
How to implement the Game of Life using numpy arrays?

In [82]:
Z = np.random.randint(0,2,(10,10))

for i in range(5): 
    # iterate 5 steps
    # Count neighbours
    N = (Z[0:-2,0:-2] + Z[0:-2,1:-1] + Z[0:-2,2:] +
         Z[1:-1,0:-2]                + Z[1:-1,2:] +
         Z[2:  ,0:-2] + Z[2:  ,1:-1] + Z[2:  ,2:])

    # Apply rules
    birth = (N==3) & (Z[1:-1,1:-1]==0)
    survive = ((N==2) | (N==3)) & (Z[1:-1,1:-1]==1)
    
    # Update Z in place
    Z[...] = 0
    Z[1:-1,1:-1][birth | survive] = 1
    
print("State after 5 iterations:\n", Z)

State after 5 iterations:
 [[0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]]


### Explanation
We calculate the number of neighbours `N` for each cell (excluding the borders to avoid boundary checks). Then we apply the Game of Life rules:
1. Birth: A dead cell with exactly 3 neighbours becomes live.
2. Survival: A live cell with 2 or 3 neighbours stays live.
We update the grid in place.

In [ ]:
State after 5 iterations:
 [[0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 1 0 0 0]
 [0 0 0 1 0 0 1 1 0 0]
 [0 0 0 1 0 1 1 0 0 0]
 [0 0 0 1 1 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]]

## Question 89
How to get the n largest values of an array?

In [83]:
Z = np.arange(10)
np.random.shuffle(Z)
n = 3

print("Array:", Z)
# Sort and take last n
print("Largest n (sort):", np.sort(Z)[-n:])

# Faster: partition
# This puts the n largest elements at the end, but not necessarily sorted
print("Largest n (partition):", np.partition(Z, -n)[-n:])

Array: [2 1 0 9 4 7 5 6 8 3]
Largest n (sort): [7 8 9]
Largest n (partition): [7 8 9]


### Explanation
We can find the `n` largest values by sorting the entire array and taking the last `n` elements. For large arrays, `np.partition(Z, -n)` is faster as it only ensures that the `n` largest elements are in the last positions, without sorting the entire array.

In [ ]:
Array: [2 8 4 9 0 6 3 1 5 7]
Largest n (sort): [7 8 9]
Largest n (partition): [7 8 9]

## Question 90
Given an arbitrary number of vectors, build the cartesian product (every combination of every item).

In [84]:
arrays = ([1, 2, 3], [4, 5], [6, 7])

# Convert to numpy arrays
arr_list = [np.asarray(a) for a in arrays]
shape = (len(x) for x in arr_list)

# Create indices grid
ix = np.indices(shape, dtype=int)
ix = ix.reshape(len(arr_list), -1).T

# Map indices to values
for n, arr in enumerate(arr_list):
    ix[:, n] = arr_list[n][ix[:, n]]

print(ix)

[[1 4 6]
 [1 4 7]
 [1 5 6]
 [1 5 7]
 [2 4 6]
 [2 4 7]
 [2 5 6]
 [2 5 7]
 [3 4 6]
 [3 4 7]
 [3 5 6]
 [3 5 7]]


### Explanation
`np.indices` generates a grid of indices for the dimensions of the input arrays. We reshape this grid to get a list of index combinations. Then, we use these indices to map back to the values in the original arrays.
Basically, if we have arrays of lengths N1, N2, N3... the result will have N1*N2*N3... rows.

In [ ]:
[[1 4 6]
 [1 4 7]
 [1 5 6]
 [1 5 7]
 [2 4 6]
 [2 4 7]
 [2 5 6]
 [2 5 7]
 [3 4 6]
 [3 4 7]
 [3 5 6]
 [3 5 7]]

## Question 91
How to create a record array from a regular array?

In [85]:
Z = np.array([("Hello", 2.5, 3),
              ("World", 3.6, 2)])
R = np.core.records.fromarrays(Z.T, 
                               names='col1, col2, col3',
                               formats = 'S8, f8, i8')
print(R)

[(b'Hello', 2.5, 3) (b'World', 3.6, 2)]


C:\Users\DKS\AppData\Local\Temp\ipykernel_21200\3979488781.py:3: DeprecationWarning: numpy.core is deprecated and has been renamed to numpy._core. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.records.
  R = np.core.records.fromarrays(Z.T,


### Explanation
We use `np.core.records.fromarrays`. Since this function expects a list of arrays (one for each column), we transpose our input array `Z.T`. We also specify the `names` and `formats` for the fields in the record array.

In [ ]:
[(b'Hello', 2.5, 3) (b'World', 3.6, 2)]

## Question 92
Consider a large vector `Z`, compute `Z` to the power of 3 using 3 different methods.

In [86]:
x = np.random.rand(5)

# Method 1
print("Method 1:", np.power(x, 3))

# Method 2
print("Method 2:", x*x*x)

# Method 3
print("Method 3:", np.einsum('i,i,i->i', x, x, x))

Method 1: [0.48922644 0.21739043 0.47503403 0.01587776 0.00215179]
Method 2: [0.48922644 0.21739043 0.47503403 0.01587776 0.00215179]
Method 3: [0.48922644 0.21739043 0.47503403 0.01587776 0.00215179]


### Explanation
We can use `np.power`, simple multiplication `*`, or `np.einsum` to compute the element-wise power. Simple multiplication is often the fastest for small integer powers.

In [ ]:
Method 1: [0.00517525 0.20349607 0.00443216 0.14441267 0.0151121 ]
Method 2: [0.00517525 0.20349607 0.00443216 0.14441267 0.0151121 ]
Method 3: [0.00517525 0.20349607 0.00443216 0.14441267 0.0151121 ]

## Question 93
Consider two arrays `A` and `B` of shape `(8,3)` and `(2,3)`. How to find rows of `A` that contain elements of each row of `B` regardless of the order of the elements in `B`?

In [87]:
A = np.random.randint(0, 5, (8, 3))
B = np.random.randint(0, 5, (2, 3))

# Since order doesn't matter, sorting is a key step
C = (A[..., np.newaxis, np.newaxis] == B)
rows = np.where(C.any((3, 1)).all(1))[0]
print("Rows of A matching any row in B (as a set):\n", rows)

Rows of A matching any row in B (as a set):
 [1 2 6 7]


### Explanation
This solution checks if elements of B are present in A. Since the problem asks for "elements of each row of B regardless of order", essentially we are treating the rows as sets.
However, a simpler way to think about "order-independent" equality of rows is to sort the rows of both A and B, then check for exact matches.
The provided solution `C.any((3,1)).all(1)` is checking if *all* elements of a row in B are present in a row in A (ignoring duplicates and order for full containment).
Let's stick to the interpretation: find rows in A that are permutations of rows in B.
`A` and `B` rows sorted:
`A_sorted = np.sort(A, axis=1)`
`B_sorted = np.sort(B, axis=1)`
Then verify if `A_sorted` equals any row of `B_sorted`.
We can use broadcasting: `(A_sorted[:,None,:] == B_sorted).all(-1).any(-1)`.

In [ ]:
Rows of A matching any row in B (as a set):
 [1 3 4]

## Question 94
Considering a 10x3 matrix, extract rows with unequal values (e.g. `[2,2,3]`).

In [88]:
Z = np.random.randint(0, 5, (10, 3))
print("Original Z:\n", Z)

# rows with unequal values logic:
# A row has unequal values if not all values are equal.
# Or does it mean "all values are distinct"?
# "Extract rows with unequal values" usually implies rows like [2,2,3] are kept, 
# while [2,2,2] are discarded. i.e., "not all equal".

E = np.all(Z[:, 1:] == Z[:, :-1], axis=1) # True if all equal
U = Z[~E]
print("\nRows with unequal values:\n", U)

Original Z:
 [[3 2 3]
 [4 4 3]
 [0 1 4]
 [1 1 1]
 [2 0 4]
 [2 1 2]
 [0 1 1]
 [0 2 1]
 [0 4 3]
 [0 0 3]]

Rows with unequal values:
 [[3 2 3]
 [4 4 3]
 [0 1 4]
 [2 0 4]
 [2 1 2]
 [0 1 1]
 [0 2 1]
 [0 4 3]
 [0 0 3]]


**Explanation**
We first create a boolean array `E` indicating for each row whether all elements are equal. We achieve this by comparing `Z[:, 1:]` with `Z[:, :-1]`. If all adjacent elements are equal across the row, the row contains only one unique value. We then negate `E` (`~E`) to select rows where at least one value is different.

**Output**
```
Original Z:
 [[0 4 3]
 [3 0 2]
 [3 2 4]
 [1 2 4]
 [0 1 4]
 [4 3 1]
 [3 0 3]
 [3 3 4]
 [3 3 1]
 [4 3 2]]

Rows with unequal values:
 [[0 4 3]
 [3 0 2]
 [3 2 4]
 [1 2 4]
 [0 1 4]
 [4 3 1]
 [3 0 3]
 [3 3 4]
 [3 3 1]
 [4 3 2]]
```

## Question 95
Convert a vector of ints into a matrix binary representation.

In [89]:
I = np.array([0, 1, 2, 3, 15, 16, 32, 64, 128], dtype=np.uint8)
print("Original I:", I)

# Method 1 using unpackbits (for uint8)
print("\nBinary representation (using unpackbits):")
print(np.unpackbits(I[:, np.newaxis], axis=1))

# Method 2 using bitwise operations (general)
I_general = np.array([0, 1, 2, 3, 15, 16, 32, 64, 128])
# Check 8 bits
bits = 8
B = ((I_general[:, None] & (1 << np.arange(bits))) > 0).astype(int)
# By default this is little-endian (2^0 is at index 0). usually we want big-endian for display (2^7 at index 0)
print("\nBinary representation (using bitwise, big-endian):")
print(B[:, ::-1])

Original I: [  0   1   2   3  15  16  32  64 128]

Binary representation (using unpackbits):
[[0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 1]
 [0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 1 0 0 0 0]
 [0 0 1 0 0 0 0 0]
 [0 1 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]

Binary representation (using bitwise, big-endian):
[[0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 1]
 [0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 1 0 0 0 0]
 [0 0 1 0 0 0 0 0]
 [0 1 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]


**Explanation**
We demonstrate two ways to obtain a binary matrix from an integer vector.
1.  **`np.unpackbits`**: This is the most efficient and direct method but requires the input to be `uint8`. We expand dimensions with `[:, np.newaxis]` to make the input 2D, so `unpackbits` with `axis=1` produces the bits for each number horizontally.
2.  **Broadcasting with Powers of 2**: For general integers, we can check matching bits using `&` and bitwise shifts. `(1 << np.arange(8))` creates powers of two. We broadcast this against `I` and check if the bit is set (`> 0`). We reverse the columns (`[:, ::-1]`) to match the standard Big-Endian (most significant bit first) representation.

**Output**
```
Original I: [  0   1   2   3  15  16  32  64 128]

Binary representation (using unpackbits):
[[0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 1]
 [0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 1 0 0 0 0]
 [0 0 1 0 0 0 0 0]
 [0 1 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]

Binary representation (using bitwise, big-endian):
[[0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 1]
 [0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 1 0 0 0 0]
 [0 0 1 0 0 0 0 0]
 [0 1 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]
```

## Question 96
Given a two dimensional array, how to extract unique rows?

In [90]:
Z = np.random.randint(0, 2, (6, 3))
print("Original Z:\n", Z)

uZ = np.unique(Z, axis=0)
print("\nUnique rows:\n", uZ)

Original Z:
 [[0 0 1]
 [0 1 0]
 [0 0 0]
 [1 1 1]
 [0 1 0]
 [1 1 0]]

Unique rows:
 [[0 0 0]
 [0 0 1]
 [0 1 0]
 [1 1 0]
 [1 1 1]]


**Explanation**
The `np.unique` function with `axis=0` identifies and returns the unique rows of the 2D array `Z`. It effectively filters out any duplicate rows.

**Output**
```
Original Z:
 [[0 0 0]
 [0 0 0]
 [0 1 1]
 [0 0 1]
 [0 0 1]
 [0 0 0]]

Unique rows:
 [[0 0 0]
 [0 0 1]
 [0 1 1]]
```

## Question 97
Considering 2 vectors A & B, write the einsum equivalent of inner, outer, sum, and mul function.

In [91]:
A = np.random.uniform(0, 1, 10)
B = np.random.uniform(0, 1, 10)

print(f"sum(A):\n{np.sum(A)}")
print(f"einsum sum:\n{np.einsum('i->', A)}\n")

print(f"multiply(A, B) (first 5):\n{np.multiply(A, B)[:5]}")
print(f"einsum mul (first 5):\n{np.einsum('i,i->i', A, B)[:5]}\n")

print(f"inner(A, B):\n{np.inner(A, B)}")
print(f"einsum inner:\n{np.einsum('i,i', A, B)}\n")

print(f"outer(A, B) (top-left 3x3):\n{np.outer(A, B)[:3, :3]}")
print(f"einsum outer (top-left 3x3):\n{np.einsum('i,j->ij', A, B)[:3, :3]}")

sum(A):
4.562102769040395
einsum sum:
4.562102769040395

multiply(A, B) (first 5):
[0.49967435 0.04773478 0.02188339 0.13335646 0.30161442]
einsum mul (first 5):
[0.49967435 0.04773478 0.02188339 0.13335646 0.30161442]

inner(A, B):
2.288458754109632
einsum inner:
2.288458754109632

outer(A, B) (top-left 3x3):
[[0.49967435 0.52606127 0.1841513 ]
 [0.04534043 0.04773478 0.01670988]
 [0.05937817 0.06251383 0.02188339]]
einsum outer (top-left 3x3):
[[0.49967435 0.52606127 0.1841513 ]
 [0.04534043 0.04773478 0.01670988]
 [0.05937817 0.06251383 0.02188339]]


**Explanation**
`np.einsum` (Einstein summation) is a powerful function to express many multi-dimensional array operations.
- `sum(A)`: `'i->'` sums over index `i` resulting in a scalar.
- `multiply(A, B)`: `'i,i->i'` takes element-wise product (same index `i` in, same `i` out).
- `inner(A, B)`: `'i,i'` sums over index `i` (dot product for 1D).
- `outer(A, B)`: `'i,j->ij'` creates a matrix from all pairs of `i` and `j`.

**Output**
```
sum(A):
5.353626346398664
einsum sum:
5.353626346398663

multiply(A, B) (first 5):
[0.06977013 0.16465181 0.45310318 0.46010381 0.07946959]
einsum mul (first 5):
[0.06977013 0.16465181 0.45310318 0.46010381 0.07946959]

inner(A, B):
2.9885607095193585
einsum inner:
2.9885607095193594

outer(A, B) (top-left 3x3):
[[0.06977013 0.01453528 0.06938629]
 [0.79033781 0.16465181 0.78598975]
 [0.45560973 0.0949176  0.45310318]]
einsum outer (top-left 3x3):
[[0.06977013 0.01453528 0.06938629]
 [0.79033781 0.16465181 0.78598975]
 [0.45560973 0.0949176  0.45310318]]
```

## Question 98
Considering a path described by two vectors (X,Y), how to sample it using equidistant samples using interpolation?

In [92]:
# Define a path (e.g., a simple curve)
phi = np.arange(0, 10*np.pi, 0.5) # Sparse points
X = phi * np.cos(phi)
Y = phi * np.sin(phi)

# Calculate Euclidean distance between consecutive points
d = np.sqrt(np.diff(X)**2 + np.diff(Y)**2)
# Cumulative distance (arc length) from the start
r = np.zeros_like(X)
r[1:] = np.cumsum(d)

# Interpolate to get equidistant samples
# Let's say we want 50 equidistant points along the path
n_samples = 50
r_new = np.linspace(0, r.max(), n_samples)
X_new = np.interp(r_new, r, X)
Y_new = np.interp(r_new, r, Y)

print("Original number of points:", len(X))
print("New number of equidistant points:", len(X_new))
print("Sample of original X (irregular steps):\n", X[:5])
print("Sample of new X (equidistant on curve):\n", X_new[:5])

Original number of points: 63
New number of equidistant points: 50
Sample of original X (irregular steps):
 [ 0.          0.43879128  0.54030231  0.1061058  -0.83229367]
Sample of new X (equidistant on curve):
 [ 0.         -2.14739316  5.81123251  2.6797698  -6.29529255]


**Explanation**
To sample a path equidistantly, we first calculate the cumulative arc length `r` at each original point. `d` is the distance between consecutive points `(X[i], Y[i])` and `(X[i+1], Y[i+1])`. We then define a new set of arc lengths `r_new` that are linearly spaced from `0` to the total length `r[-1]`. Finally, we use `np.interp` to find the corresponding `X` and `Y` coordinates for these new arc length values.

**Output**
```
Original number of points: 63
New number of equidistant points: 50
Sample of original X (irregular steps):
 [ 0.          0.43879128  0.54030231  0.1061058  -0.83229367]
Sample of new X (equidistant on curve):
 [ 0.         -2.14739316  5.81123251  2.6797698  -6.29529255]
```

## Question 99
Given an integer n and a 2D array X, select from X the rows which can be interpreted as draws from a multinomial distribution with n degrees, i.e., the sum of row elements is n.

In [93]:
n = 4
X = np.random.randint(0, 3, (10, 3))
# Just to be sure we have at least one match
X[0] = [1, 1, 2] 

print(f"Target sum n = {n}")
print("Original X:\n", X)
row_sums = np.sum(X, axis=1)
print("Row sums:", row_sums)

M = X[row_sums == n]
print("\nRows with sum = n:\n", M)

Target sum n = 4
Original X:
 [[1 1 2]
 [1 1 2]
 [2 1 2]
 [1 0 1]
 [1 2 0]
 [0 0 0]
 [2 2 1]
 [0 2 1]
 [1 1 1]
 [0 2 2]]
Row sums: [4 4 5 2 3 0 5 3 3 4]

Rows with sum = n:
 [[1 1 2]
 [1 1 2]
 [0 2 2]]


**Explanation**
A row can be considered a draw from a multinomial distribution with $n$ degrees if the sum of its elements equals $n$. We filter the rows by calculating the sum along `axis=1` and comparing it to `n`.

**Output**
```
Target sum n = 4
Original X:
 [[1 1 2]
 [1 2 1]
 [0 2 2]
 [2 0 2]
 [1 0 0]
 [1 2 2]
 [1 1 1]
 [2 2 2]
 [1 1 0]
 [0 2 2]]
Row sums: [4 4 4 4 1 5 3 6 2 4]

Rows with sum = n:
 [[1 1 2]
 [1 2 1]
 [0 2 2]
 [2 0 2]
 [0 2 2]]
```

## Question 100
Compute bootstrapped 95% confidence intervals for the mean of a 1D array X.

In [94]:
X = np.random.randn(100) # Random data, mean roughly 0
N = len(X)
B = 1000 # Number of bootstrap samples

# Generate bootstrap samples (resampling with replacement)
# We generate B sets of indices, each of size N
idx = np.random.randint(0, N, (B, N))
# Select the data using these indices
bootstrap_samples = X[idx]

# Compute the mean for each bootstrap sample
bootstrap_means = np.mean(bootstrap_samples, axis=1)

# Compute the 2.5th and 97.5th percentiles of the bootstrap means
CI = np.percentile(bootstrap_means, [2.5, 97.5])

print(f"Original Mean: {X.mean():.4f}")
print(f"95% Confidence Interval: [{CI[0]:.4f}, {CI[1]:.4f}]")

Original Mean: 0.1637
95% Confidence Interval: [-0.0389, 0.3624]


**Explanation**
Bootstrapping is a resampling technique used to estimate statistics on a population by sampling a dataset with replacement.
1.  We generate `B` arrays of indices, each of size `N`, where each index is drawn randomly from `0` to `N-1`.
2.  We use these indices to create `B` resampled versions of `X`.
3.  We calculate the mean for each of these resampled datasets.
4.  The 95% confidence interval is determined finding the 2.5th and 97.5th percentiles of the distribution of these bootstrap means.

**Output**
```
Original Mean: -0.0037
95% Confidence Interval: [-0.2032, 0.2272]
```